File: 116. Populating Next Right Pointers in Each Node.cpp
/**
 * Definition for binary tree with next pointer.
 * struct TreeLinkNode {
 *  int val;
 *  TreeLinkNode *left, *right, *next;
 *  TreeLinkNode(int x) : val(x), left(NULL), right(NULL), next(NULL) {}
 * };
 */
class Solution {
public:
    // BFS: O(N)time & space, 40.61%, ok
    void connect(TreeLinkNode *root) {
        if(!root) return;
        queue<TreeLinkNode*> q;
        q.push(root);
        while(q.size()){
            int size = q.size();
            TreeLinkNode* prev = NULL;
            while(size--){
                TreeLinkNode* temp = q.front();
                q.pop();
                if(prev) prev->next = temp;
                prev = temp;
                if(temp->left) q.push(temp->left);
                if(temp->right) q.push(temp->right);
            }
        }
    }

    // DFS on the left child: O(N)time, O(1)space, 40.61%, ok
    void connect(TreeLinkNode *root) {
        if(!root) return;
        while(root->left){
            TreeLinkNode* p = root;
            while(p){
                p->left->next = p->right;
                if(p->next) p->right->next = p->next->left;
                p = p->next;
            }
            root = root->left;
        }
    }
};

File: 116. Populating Next Right Pointers in Each Node.py
from collections import deque
# Definition for binary tree with next pointer.
# class TreeLinkNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
#         self.next = None

class Solution:
    # BFS: O(N) time, O(N) space
    # @param root, a tree link node
    # @return nothing
    def connect(self, root):
        if root is None:
            return 
        queue, prev = deque([root]), None
        while queue:
            levelSize = len(queue)
            prev = None
            while levelSize:
                node = queue.popleft()
                node.next = prev
                prev = node
                queue += filter(None, [node.right, node.left])
                levelSize -= 1
    
    # !! DFS: O(N) time, O(1) space
    def connect(self, root):
        prev, curr = root, None
        while prev:
            curr = prev
            while curr:
                if curr.left: 
                    curr.left.next = curr.right
                if curr.right and curr.next:
                    curr.right.next = curr.next.left
                curr = curr.next
            prev = prev.left
        
        
      
        